### DISCUSSION TITLE
how to release memory for C++ VImage

### DISCUSSION THREAD

**Discussion by oldercoder** [libvips/libvips#2920] [Help wanted]
Hi @jcupitt ,
I read lots of your answers/discussion, that's help me a lot on how to use libvips correctly, thanks!

Now, there is a issue drag me crazy weeks, the senario is :
I need process many tiff files(500+) within a for loop, in each loop I will read each file from disk and do heavy actions/computations like extract_area, insert, embed, mapim, pow, log etc., then save the image to disk. The symptom is, after process each tiff file, the memory goes up and up, never release back to OS, the whole process will crash after all the memory was eat up. 
So, my question is, in C++ code, how do I release the memory after I process one single tiff file, then goes to another? ( since all the computation are intermediate) 
I read the libVips C++ doc, "[VImage](https://www.libvips.org/API/8.11/cpp/classVImage.html) objects are smart pointers over the underlying VipsImage objects. They manage the complications of GLib's ref and unref system for you." , it seems at least for caller, there is nothing should do if you use C++ VImage object, I don't know if I understand this correctly? Or, should I still need take care of  object ref/unref, and how?

**Comment by jcupitt** [comment]
Hi @oldercoder,

It should work. If you post some code I can run I can check it.

You could try shrinking the libvips cache -- it'll keep some images around in case you need them again, and that's obviously not so useful for batch processing. Try adding `vips_cache_set_max(10);` soon after startup.

https://www.libvips.org/API/current/VipsOperation.html#vips-cache-set-max

**Comment by jcupitt** [comment]
I tried:

```C++
/* compile with
 *
 *      g++ -g -Wall try335.cpp `pkg-config vips-cpp --cflags --libs`
 */

#include <iostream>
#include <vips/vips8>

using namespace vips;
using namespace std;

int
main( int argc, char **argv )
{
        if (VIPS_INIT (argv[0]))
                vips_error_exit (NULL);

        vips_cache_set_max(10);

        try {
                for (int i = 2; i < argc; i++) {
                        printf("processing %s ...\n", argv[i]);
                        VImage image = VImage::new_from_file(argv[i]);
                        image = image.gravity(VIPS_COMPASS_DIRECTION_CENTRE, 2000, 2000, VImage::option()
                                ->set("extend", "black"));
                        image = image.pow(2.4).log();
                        image.write_to_file(argv[1], VImage::option()
                                ->set("strip", TRUE)
                                ->set("Q", 70));
                }
        }
        catch (const VError &err ) {
                cout << err.what();
        }

        return 0;
}
```

On 1000 1450 x 2048 TIFFs and it ran in under 2gb of memory. 

You don't need random access for stuff like this, so you could leave the cache alone and instead use:

```C++
                        VImage image = VImage::new_from_file(argv[i], VImage::option()
                                ->set("access", "sequential"));
```

It'll run in about 700MB of memory.

This PC has 32 threads. If I limit the concurrency, it's a bit slower, but runs in 100MB or so:

```
VIPS_CONCURRENCY=6 ../a.out y.tif *.tif
```

**Reply by oldercoder** [reply]
Thanks for your testing code, I think I found the memory leak. It's all about a transform code from VIPS image to armadillo matrix, I forgot g_free the buffer after write_to_memory :-).

Thanks again for your kindly guidance.
